### PR TITLE
MCP-1969 End-to-end test phase 1

### PR DESCRIPTION
**Pull request by au2**
## What was the problem?

There was no end to end test which was similar to what QA team performing.  With libraries changing frequently we need one.

There was no end to end test.

Associated tickets or Slack threads:

- [MCP-1969](https://amida.atlassian.net/browse/MCP-1969)

## How does this fix it?[^1]

This adds a new gradle task ":app:end2endTest". This task assumes the app docker-compose has already been called and all services are running locally. Based on gold files this test calls rest end points to generate health assessment and evidence pdf, and verifies the results.

Currently this test is intended to run only locally. Once this PR is cleared the test will be put in github workflows. That is the next step.

Notes

1) This depends on on two existing records on Lighthouse Health API sandbox. Future versions might use a mock api to avoid failures due to changes to those patients or filtering changes based on dates (ex: this test will fail in July since one of the blood pressure readings is in June).
2) Evidence PDF data fields are only sanity checked with counts. Future versions might use a more sophisticated parser to extract everything.
3) No negative or break tests yet. Those are for future.

## How to test this PR
- Make sure all the services are running ex "./gradlew resetloop".
- Run "./gradlew end2endTest" and make sure it passes. 
- You can change various things in the gold files to make sure it fails. For example change a bp value or remove a medication.

[^1]: [Pull-Requests guidelines](https://github.com/department-of-veterans-affairs/abd-vro/wiki/Pull-Requests). If PR is significant, update [Current Software State](https://github.com/department-of-veterans-affairs/abd-vro/wiki/Current-Software-State) wiki page.
[^secrel]: To check if a PR will succeed in the SecRel workflow, [test PRs in the SecRel pipeline](https://github.com/department-of-veterans-affairs/abd-vro-internal/wiki/Secure-Release-process#to-test-prs-in-the-secrel-pipeline).

### CODE REVIEW

**Review by wlin2019 — APPROVED** [review]
LGTM.

File: app/src/end2endTest/resources/test-6602-01/veteranInfo.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+  "first": "Mary",
+  "middle": "Y",
+  "last": "Buck",
+  "suffix": "Jr",
+  "birthdate": "05/23/1975"
+}
```
Comment (dimitri-amida) [inline app/src/end2endTest/resources/test-6602-01/veteranInfo.json:7]: newline in the end

File: app/src/end2endTest/resources/test-6602-01/assessment.json [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+        ],
+        "route": "As directed by physician."
+      }
+    ]
+  },
+  "evidenceSummary": {
+    "relevantMedCount": 4,
+    "totalMedCount": 11
+  }
+}
```
Comment (dimitri-amida) [inline app/src/end2endTest/resources/test-6602-01/assessment.json:153]: newline in the end

File: app/src/end2endTest/resources/test-7101-01/veteranInfo.json [hunk 3]
```diff
@@ -0,0 +1,7 @@
+{
+  "first": "Joe",
+  "middle": "M",
+  "last": "Doe",
+  "suffix": "Jr",
+  "birthdate": "01/04/1971"
+}
```
Comment (dimitri-amida) [inline app/src/end2endTest/resources/test-7101-01/veteranInfo.json:7]: newline

**Review by dimitri-amida — COMMENTED** [review]
I checkout out this branch, and the end2endTest folder does not appear as an active folder in Intellij. How can I run this test?

File: app/src/end2endTest/java/gov/va/vro/end2end/util/RestHelper.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+@Getter
+@Setter
+public class RestHelper {
+  private static final long DEFAULT_TIMEOUT = 15000;
+  private String apiKey;
+
+  private WebTestClient buildClient() {
+    return WebTestClient.bindToServer()
+        .baseUrl("http://localhost:8080/v1")
```
[the remark this answers is not in the document]
Comment (au2) [reply app/src/end2endTest/java/gov/va/vro/end2end/util/RestHelper.java]: done.

File: app/src/end2endTest/java/gov/va/vro/end2end/util/PdfText.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+ * sanity checks for now until a more sophisticated parser extracts all data.
+ */
+public class PdfText {
+  private String pdfText;
+
+  public PdfText(String pdfText) {
+    this.pdfText = pdfText;
+  }
+
+  private int findOccurrence(String target) {
```
[the remark this answers is not in the document]
Comment (au2) [reply app/src/end2endTest/java/gov/va/vro/end2end/util/PdfText.java:20]: Not really. Test data should be erroneous for that to happen for the current existing gold files. And the test should fail for those cases.

File: app/src/end2endTest/resources/test-6602-01/veteranInfo.json (hunk 1 above)
Comment (au2) [reply app/src/end2endTest/resources/test-6602-01/veteranInfo.json]: done

File: app/src/end2endTest/resources/test-6602-01/assessment.json (hunk 2 above)
Comment (au2) [reply app/src/end2endTest/resources/test-6602-01/assessment.json]: done

File: app/src/end2endTest/resources/test-7101-01/veteranInfo.json (hunk 3 above)
Comment (au2) [reply app/src/end2endTest/resources/test-7101-01/veteranInfo.json]: done

**Review by dimitri-amida — APPROVED** [review]
These are great!

**Comment by sethdarragile6** [comment]
Sorry if this is a bad form to ask here, but I'm still dazzled by which slack channel to ask what where, at the moment. When I run `./gradlew :app:end2endTest`, the 2 tests fail because of 404s. Based on what I've seen before, I imagine that that's mostly because the **X-API-Key** is blank when calling out. What's the best way to inject that in from the command line? Or is there docker-compose override I can use? Thanks...

image

**Pull request merged**